File: java/nitf/src/main/java/nitf/GraphicSubheader.java
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 * 
 * (C) Copyright 2004 - 2008, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public 
 * License along with this program; if not, If not, 
 * see <http://www.gnu.org/licenses/>.
 *
 */

package nitf;

import java.io.PrintStream;

/**
 * A representation of the NITF Graphic (Symobl) subheader
 */
public final class GraphicSubheader extends NITFObject
{

    /**
     * @see NITFObject#NITFObject(long)
     */
    GraphicSubheader(long address)
    {
        super(address);
    }

    /**
     * Returns the File Part Type (SY) Field
     *
     * @return the File Part Type (SY) Field
     */
    public native synchronized Field getFilePartType();

    /**
     * Returns the Graphic Identifer (SID) Field
     *
     * @return the Graphic Identifer (SID) Field
     */
    public native synchronized Field getGraphicID();

    /**
     * Returns the Graphic Name (SNAME) Field
     *
     * @return the Graphic Name (SNAME) Field
     */
    public native synchronized Field getName();

    /**
     * Returns the Graphic Security Classification (SSCLAS) Field
     *
     * @return the Graphic Security Classification (SSCLAS) Field
     */
    public native synchronized Field getSecurityClass();

    /**
     * Returns the FileSecurity associated with this Graphic Header
     *
     * @return the FileSecurity associated with this Graphic Header
     */
    public native synchronized FileSecurity getSecurityGroup();

    /**
     * Returns the Encryption (ENCRYP) Field
     *
     * @return the Encryption (ENCRYP) Field
     */
    public native synchronized Field getEncrypted();

    /**
     * Returns the Graphic Type (SFMT) Field
     *
     * @return the Graphic Type (SFMT) Field
     */
    public native synchronized Field getStype();

    /**
     * Returns the Reserved for future use (SSTRUCT) Field
     *
     * @return the Reserved for future use (SSTRUCT) Field
     */
    public native synchronized Field getRes1();

    /**
     * Returns the Graphic Display Level (SDLVL) Field
     *
     * @return the Graphic Display Level (SDLVL) Field
     */
    public native synchronized Field getDisplayLevel();

    /**
     * Returns the Graphic Attachment Level (SALVL) Field
     *
     * @return the Graphic Attachment Level (SALVL) Field
     */
    public native synchronized Field getAttachmentLevel();

    /**
     * Returns the Graphic Location - (SLOC) Field
     *
     * @return the Graphic Location - (SLOC) Field
     */
    public native synchronized Field getLocation();

    /**
     * Returns the First Graphic Bound Location - (SBND1) Field
     *
     * @return the First Graphic Bound Location - (SBND1) Field
     */
    public native synchronized Field getBound1Loc();

    /**
     * Returns the Graphic Color (SCOLOR) Field
     *
     * @return the Graphic Color (SCOLOR) Field
     */
    public native synchronized Field getColor();

    /**
     * Returns the Second Graphic Bound Location - (SBND2) Field
     *
     * @return the Second Graphic Bound Location - (SBND2) Field
     */
    public native synchronized Field getBound2Loc();

    /**
     * Returns the Reserved For Future Use (SRES2) Field
     *
     * @return the Reserved For Future Use (SRES2) Field
     */
    public native synchronized Field getRes2();

    /**
     * Returns the Graphic Extended Subheader Data Length (SXSHDL) Field
     *
     * @return the Graphic Extended Subheader Data Length (SXSHDL) Field
     */
    public native synchronized Field getExtendedHeaderLength();

    /**
     * Returns the Graphic Extended Subheader Overflow (SXSOFL) Field
     *
     * @return the Graphic Extended Subheader Overflow (SXSOFL) Field
     */
    public native synchronized Field getExtendedHeaderOverflow();

    /**
     * Returns the Graphic Extended Subheader Data Extensions segment
     *
     * @return the Graphic Extended Subheader Data Extensions segment
     */
    public native synchronized Extensions getExtendedSection();


    /**
     * Prints the data associated with the GraphicSubheader to a PrintStream
     *
     * @param out
     */
    public void print(PrintStream out)
    {
        out.println("FilePartType = [" + getFilePartType() + "]");
        out.println("GraphicID = [" + getGraphicID() + "]");
        out.println("GraphicName = [" + getName() + "]");
        out.println("GraphicSecurityClass = [" + getSecurityClass() + "]");

        getSecurityGroup().print(out);

        out.println("Encryption = [" + getEncrypted() + "]");
        out.println("GraphicType = [" + getStype() + "]");
        out.println("Reserved1 = [" + getRes1() + "]");
        out.println("DisplayLevel = [" + getDisplayLevel() + "]");
        out.println("AttachmentLevel = [" + getAttachmentLevel() + "]");
        out.println("GraphicLocation = [" + getLocation() + "]");
        out.println("BoundLocation1 = [" + getBound1Loc() + "]");
        out.println("Color = [" + getColor() + "]");
        out.println("BoundLocation2 = [" + getBound2Loc() + "]");
        out.println("Reserved2 = [" + getRes2() + "]");
        out.println("Reserved2 = [" + getExtendedHeaderLength() + "]");
    }
}

File: c++/nitf/source/GraphicSubheader.cpp
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 *
 * (C) Copyright 2004 - 2010, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, If not,
 * see <http://www.gnu.org/licenses/>.
 *
 */

#include "nitf/GraphicSubheader.hpp"

using namespace nitf;

GraphicSubheader::GraphicSubheader(const GraphicSubheader & x)
{
    setNative(x.getNative());
}


GraphicSubheader & GraphicSubheader::operator=(const GraphicSubheader & x)
{
    if (&x != this)
        setNative(x.getNative());
    return *this;
}

GraphicSubheader::GraphicSubheader(nitf_GraphicSubheader * x)
{
    setNative(x);
    getNativeOrThrow();
}

GraphicSubheader::GraphicSubheader() throw(nitf::NITFException)
{
    setNative(nitf_GraphicSubheader_construct(&error));
    getNativeOrThrow();
    setManaged(false);
}


nitf::GraphicSubheader GraphicSubheader::clone() throw(nitf::NITFException)
{
    nitf::GraphicSubheader dolly(
        nitf_GraphicSubheader_clone(getNativeOrThrow(), &error));
    dolly.setManaged(false);
    return dolly;
}

GraphicSubheader::~GraphicSubheader(){}

nitf::Field GraphicSubheader::getFilePartType()
{
    return nitf::Field(getNativeOrThrow()->filePartType);
}

nitf::Field GraphicSubheader::getGraphicID()
{
    return nitf::Field(getNativeOrThrow()->graphicID);
}

nitf::Field GraphicSubheader::getName()
{
    return nitf::Field(getNativeOrThrow()->name);
}

nitf::Field GraphicSubheader::getSecurityClass()
{
    return nitf::Field(getNativeOrThrow()->securityClass);
}

nitf::FileSecurity GraphicSubheader::getSecurityGroup()
{
    return nitf::FileSecurity(getNativeOrThrow()->securityGroup);
}

void GraphicSubheader::setSecurityGroup(nitf::FileSecurity value)
{
    //release the owned security group
    nitf::FileSecurity fs = nitf::FileSecurity(getNativeOrThrow()->securityGroup);
    fs.setManaged(false);

    //have the library manage the new securitygroup
    getNativeOrThrow()->securityGroup = value.getNative();
    value.setManaged(true);
}


nitf::Field GraphicSubheader::getEncrypted()
{
    return nitf::Field(getNativeOrThrow()->encrypted);
}

nitf::Field GraphicSubheader::getStype()
{
    return nitf::Field(getNativeOrThrow()->stype);
}

nitf::Field GraphicSubheader::getRes1()
{
    return nitf::Field(getNativeOrThrow()->res1);
}

nitf::Field GraphicSubheader::getDisplayLevel()
{
    return nitf::Field(getNativeOrThrow()->displayLevel);
}

nitf::Field GraphicSubheader::getAttachmentLevel()
{
    return nitf::Field(getNativeOrThrow()->attachmentLevel);
}

nitf::Field GraphicSubheader::getLocation()
{
    return nitf::Field(getNativeOrThrow()->location);
}

nitf::Field GraphicSubheader::getBound1Loc()
{
    return nitf::Field(getNativeOrThrow()->bound1Loc);
}

nitf::Field GraphicSubheader::getColor()
{
    return nitf::Field(getNativeOrThrow()->color);
}

nitf::Field GraphicSubheader::getBound2Loc()
{
    return nitf::Field(getNativeOrThrow()->bound2Loc);
}

nitf::Field GraphicSubheader::getRes2()
{
    return nitf::Field(getNativeOrThrow()->res2);
}

nitf::Field GraphicSubheader::getExtendedHeaderLength()
{
    return nitf::Field(getNativeOrThrow()->extendedHeaderLength);
}

nitf::Field GraphicSubheader::getExtendedHeaderOverflow()
{
    return nitf::Field(getNativeOrThrow()->extendedHeaderOverflow);
}

nitf::Extensions GraphicSubheader::getExtendedSection()
{
    return nitf::Extensions(getNativeOrThrow()->extendedSection);
}

void GraphicSubheader::setExtendedSection(nitf::Extensions value)
{
    if (getNativeOrThrow()->extendedSection)
    {
        //release the one currently "owned", if different
        nitf::Extensions exts = nitf::Extensions(getNativeOrThrow()->extendedSection);
        if (exts != value)
            exts.setManaged(false);
    }

    //have the library manage the "new" one
    getNativeOrThrow()->extendedSection = value.getNative();
    value.setManaged(true);
}
